### PR TITLE
Analysis/in datatypes

### PR DESCRIPTION
**Pull request by uxmal**
This PR uses a variant of the `ExpressionTypeDescender` to extract type information from local procedures. This helps us eliminate more `fp-XX` expressions and replace them with `&tLocXX`.

### CODE REVIEW

**Review by ptomin — CHANGES_REQUESTED** [review]
I have added some notes.

**Comment by ptomin** [comment]
One more note: it still not work for https://github.com/uxmal/reko/blame/master/subjects/PE/x86/LocalStackVariables sample

File: src/Decompiler/Typing/ExpressionTypeDescender.cs [hunk 1]
```diff
@@ -407,13 +409,13 @@ private DataType PushMinuendDataType(DataType dtDiff, DataType dtSub)
             {
                 if ((dtSub.Domain & Domain.Integer) != 0)
                     return PrimitiveType.Create(Domain.Pointer, dtDiff.BitSize);
-                throw new TypeInferenceException($"Not handling {dtDiff} and {dtSub} yet.");
+                //throw new TypeInferenceException($"Not handling {dtDiff} and {dtSub} yet.");
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Typing/ExpressionTypeDescender.cs:412]: Done. Note how the new `DataType.Domain` property pays off in simplifying the code.

File: src/Decompiler/Typing/ExpressionTypeDescender.cs [hunk 2]
```diff
@@ -424,22 +426,23 @@ private DataType PushSubtrahendDataType(DataType dtDiff, DataType dtMin)
             {
                 if (dtMin.Domain == Domain.Pointer)
                     return PrimitiveType.Create(Domain.Integer, dtDiff.BitSize);
-                throw new TypeInferenceException(string.Format("Not handling {0} and {1} yet", dtDiff, dtMin));
+                //throw new TypeInferenceException(string.Format("Not handling {0} and {1} yet", dtDiff, dtMin));
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Decompiler/Typing/ExpressionTypeDescender.cs:429]: Done.

File: src/Core/Expressions/PhiFunction.cs [hunk 3]
```diff
@@ -80,5 +80,11 @@ public PhiArgument(Block block, Expression value)
             this.Value = value;
             this.Block = block;
         }
+
+        public void Deconstruct(out Block block, out Expression value)
+        {
+            block = this.Block;
+            value = this.Value;
+        }
```
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Core/Expressions/PhiFunction.cs:88]: Actually, it's used in places where the `PhiArgument` is deconstructed. For instance in the code starting on line 52 in `LocalTypeDescender.cs`:
```C#
                    var tv = TypeVar(phi.Dst);
                    foreach (var (_, v) in phi.Src.Arguments)
                    {
                        store.MergeClasses(tv, TypeVar(v));
                    }
```
The `(_, v)` is deconstructing the `PhiArgument` by calling the `Deconstruct` method.

[the remark this answers is not in the document]
Comment (ptomin) [reply src/Core/Expressions/PhiFunction.cs:88]: Oh, `Deconstruct` is magic method. I did not know that.

**Comment by uxmal** [comment]
@ptomin : I looked at `LocalStackVariables`. The machine code is 
```

00401022 lea eax,[ebp-18h]
00401025 push eax
00401026 lea ecx,[ebp-28h]
00401029 push ecx
0040102A call GetMin
```
which gets rewritten, SSA converted and value propagated to:
```
	// res = GetMin(&s1, &s2)
	eax_20 = fp - 28<i32>
	esp_21 = fp - 0x30<32>
	Mem22[fp - 0x30<32>:word32] = fp - 28<i32>
	ecx_23 = fp - 44<i32>
	esp_24 = fp - 0x34<32>
	Mem25[fp - 0x34<32>:word32] = fp - 44<i32>
	call GetMin (retsize: 4;)
		uses: Stack +0004:Mem25[fp - 0x34<32>:word32],Stack +0008:Mem25[fp - 0x30<32>:word32]
		defs: C:C_31,eax:eax_26,ecx:ecx_27,edx:edx_28,FPUF:FPUF_29,O:O_33,S:S_30,Z:Z_32
```
Value propagation has not occurred on the arguments to `GetMin` because they are not registers/identifiers; they are in the form of `Memxx[fp - XX]`. The logic in `EscapedFrameIntervalsFinder.IsFrameAccess` expect arguments to be of the form `fp-XX`. 

A future commit will have to address this by walking the stack backwards from the `call` to detect arguments pushed on the stack and resolve them that way.

File: subjects/Elf/ARM/angr-685/RTOSDemo.reko/RTOSDemo_text.c [hunk 4]
```diff
@@ -123,30 +123,39 @@ ui32 g_dw80F8 = 0x8069; // 000080F8
...
+	tLoc15.dw000D = r6;
+	tLoc15.t0011 = lr;
 	struct Eq_n * r5_n = g_ptr8174;
+	tLoc15.dw0001 = 0x00;
 	Eq_n r0_n = UARTIntStatus(r5_n, 0x01);
 	UARTIntClear(r5_n, r0_n);
 	if (r0_n << 27 < 0x00 && *g_ptr8178 << 25 < 0x00)
+	{
+		tLoc15.t0000 = r5_n->t0000;
 		xQueueGenericSendFromISR(fp - 0x15, r0_n, fp - 20, 0x00, lr, cpsr);
```
[the remark this answers is not in the document]
Comment (uxmal) [reply subjects/Elf/ARM/angr-685/RTOSDemo.reko/RTOSDemo_text.c:142]: Fixed on my machine.

File: src/Core/Expressions/PhiFunction.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (uxmal) [reply src/Core/Expressions/PhiFunction.cs:88]: C# is a growing language. Soon it will be as huge as C++ :)

**Comment by ptomin** [comment]
`EscapedFrameIntervalsFinder` designed to make possible to do such transformations. `EscapedFrameIntervalsFinder.Context` (implementation of ` EvaluationContext`) should do its job properly even for Memory Accesses propagation. So walking the stack backwards is not required.
Are you sure that data types of `GetMin` call parameters discovered correctly by `UserRegisterFinder`/` LocalTypeDescender`?

**Comment by uxmal** [comment]
@ptomin : this is starting to look pretty good.

File: subjects/PE/x86/LocalStackVariables/LocalStackVariables.reko/LocalStackVariables_text.c [hunk 5]
```diff
@@ -7,32 +7,38 @@
...
+	Eq_n tLoc2C;
+	Eq_n tLoc1C;
 	// s1.i = 0
+	tLoc2C.dw0000 = 0x00;
 	// s1.d = 1.0
+	tLoc2C.r0008 = 1.0;
 	// s2.i = 10
+	tLoc1C.dw0000 = 0x0A;
 	// s2.d = 11.0
+	tLoc1C.r0008 = 11.0;
```
Comment (ptomin) [inline subjects/PE/x86/LocalStackVariables/LocalStackVariables.reko/LocalStackVariables_text.c:19]: Nice output!

**Review by ptomin — APPROVED** [review]
Looks good, now.

**Pull request closed without merging**